### PR TITLE
Add rich progress bar and RichWriter support

### PR DESCRIPTION
**Pull request by Ovsyanka83**
This is still somewhat of a rough draft because the progress bar will be visible even if the out file has been specified. But if the idea itself is accepted, I can fix all the issues that you find here.

Issue link:
#73

### CODE REVIEW

File: pygount/common.py
Comment (roskakori) [inline pygount/common.py:13]: This should be 1.3.0.

**Review by roskakori — CHANGES_REQUESTED** [review]

**Comment by roskakori** [comment]
The 3.7 build fails because your fork cannot access the coveralls token.

You can ignore this, for you the 3.9 build is the essential one.

File: pygount/common.py
Comment (roskakori) [reply pygount/common.py:13]: If you merge the current master, `__version__` is now managed by poetry. You will probably have a merge conflict here that can be resolved by accepting the changes on master and discarding your variant of this line.

Comment (Ovsyanka83) [reply pygount/common.py:13]: Oh. I love that you switched to poetry. I'll merge everything nicely within the next 1-2 days.

File: pygount/command.py
Comment (roskakori) [inline pygount/command.py:340]: Is there any reason only the `SummaryWriter` would track the progress? To me it seems a progress bar would be nice for all writers.

File: pygount/write.py
Comment (roskakori) [inline pygount/write.py:146]: I think this would be easier to read and understand:
```python
_COLUMNS_WITH_JUSTIFY = (
   ("Language", "left"),
   ("Files", "right"),
   ("Blank", "right"),
   ("Comment", "right"),
   ("Code", "right"),
)
...

table = Table()
for column, justify in _COLUMNS_WITH_JUSTIFY:
   table.add(column, justify)
```
The code above also spells out the unnecessary abbreviation `col` and reverts the column headings to consistently start with an upper case letter.

File: pygount/write.py
Comment (roskakori) [inline pygount/write.py:155]: Would a more pythonic list comprehension work here, too? AFAIK the `map` function is kind of an relic from the time when there was no list comprehension.

Something like:
```python
[str(value) for value in (
    language_summary.file_count,
    language_summary.empty_count,
    language_summary.documentation_count,
    language_summary.code_count,
)]
```

File: pygount/command.py
Comment (roskakori) [inline pygount/command.py:341]: If I read [the progress documentation](https://rich.readthedocs.io/en/stable/progress.html) correctly, this tracks progress by number of files processed.

I suppose it's fine for an initial implementation. Let me know if I should look into making a more ambitious tracking based on file size (after a merge).

**Review by roskakori — CHANGES_REQUESTED** [review]
Looks good, found a few minor issues.

Very nice that so much of the old code to render the table is gone.

File: pygount/command.py
Comment (Ovsyanka83) [reply pygount/command.py:340]: Only the summary writer displays the results at the end. I can tinker a little bit and add it for everything. The original reason I used this check is to leave your project completely untouched unless rich was used. So yeah, I will fix that.

File: pygount/write.py
Comment (Ovsyanka83) [reply pygount/write.py:155]: Actually, map is not exactly a relic. It's just another method of doing such things in functional programming. Also, when you're not using lambdas, map is roughly 20%-40% faster than listcomp so I use it in all simple cases. However, if you prefer listcomps for such cases, we can use it -- the table creation is not necessary to optimize anyways.

File: pygount/command.py
Comment (Ovsyanka83) [reply pygount/command.py:341]: We should do a small test that times the original execution vs execution with prior file size check. It felt to my like an extremely expensive operation but after using pygount for a while, I have realized just how slow pygments can get so we can try implementing file size tracking even before the merge. I will work on this tomorrow.

File: pygount/write.py
Comment (roskakori) [reply pygount/write.py:155]: I found [the old discussion](https://www.artima.com/weblogs/viewpost.jsp?thread=98196) where Guido was thinking about removing `map()`, but it managed to survive this. But I agree with his point that list comprehension is easier to read. So if you don't mind I'd prefer it over `map()` here.

Good to know that there is a performance difference, thanks for enlightening me.

File: pygount/command.py
Comment (roskakori) [reply pygount/command.py:341]: I've been thinking a bit more about this and I believe for the time being we should leave it at the count based progress.

The main reason being files that can be rather huge but processed very quickly, especially generated and duplicate files. For a decent progress, there would probably need to be two passes:
1. Detect and process files that do not have to be lexed by pygments
2. Process files that are lexed by pygments.

For a total progress, these two passes would need to be accounted for by some heuristic, like pass 1 = 10% and pass 2 = 90%. And the progress within pass 1 would be count based, within pass 2 size based.

But I'd rather we move this two a separate ticked I and look at it on my own (unless you really want to).

**Comment by roskakori** [comment]
@Ovsyanka83 The existing tests for the `SummaryWriter` are failing. Can you take a look please?

**Comment by Ovsyanka83** [comment]
I feel like the "size" test for summary needs to be either rewritten or deleted. It is now handled by rich in its entirety so we don't have to worry about it. Also, the test fails because rich folds lines for readable output, while the old version didn't.

File: pygount/command.py
Comment (Ovsyanka83) [inline pygount/command.py:333]: Progress is not down there with the other context managers because the code right below it references sys.stdout.

Thing is: rich makes progress appear below all printed text by replacing sys.stdout with its wrapper during the Progress.__enter__. However, if we make a reference to sys.stdout before __enter__, the progress bar display will be broken.

But because it's a whole another level of indentation, I decided to refactor the code below which is why we have so many changes. It is **almost** equivalent. Except for the fact that we don't raise our own OSError.

Comment (roskakori) [reply pygount/command.py:333]: Makes sense.

Didn't know about `contextlib.nullcontext()`, learned something new. 🙂 

I'm not so sure about removing rid of the `OSError`. After the merge I'm going to this to ensure in case of write errors we still get a meaningful error message.

File: pygount/common.py
Comment (roskakori) [reply pygount/common.py]: Resolved with move to poetry.

File: pygount/write.py
Comment (roskakori) [reply pygount/write.py]: Cleaned up.

**Review by roskakori — APPROVED** [review]
Even better! 👍

File: pygount/write.py
Comment (roskakori) [reply pygount/write.py]: Cleaned up.

**Comment by roskakori** [comment]
>  I have removed one of the test_write tests. I don't really know how to sanely implement it so I didn't rewrite it

Makes sense as the layout is now down by rich.

**Comment by roskakori** [comment]
@Ovsyanka83 Darn, tests fail with Python 3.6 because of:
> AttributeError: module 'contextlib' has no attribute 'nullcontext'

This [stackoverflow discussion](https://stackoverflow.com/questions/45187286/how-do-i-write-a-null-no-op-contextmanager-in-python) suggest how to implement it for 3.6.

Another solution would be to simply drop support for Python 3.6 and also remove it from the CI build. After all, Python 3.6 has reached end-of-life in September 2021. Personally I don't care much because everything I've been doing has been Python 3.8+ for a while.

However, I also know that Ubuntu is pretty popular and Ubuntu 18 LTS still uses Python 3.6 by default.

I'd say your choice.

**Comment by Ovsyanka83** [comment]
@roskakori oh, I am so sorry, I completely forgot that I was testing on 3.7 and didn't notice the docs section that stated that nullcontext is 3.7+. My laptop has some issues with pyenv + poetry at this point so I couldn't use 3.6. I definitely want to backport to 3.6 because I want to use pygount at work (we don't have access to most of the non-python software so no cloc available) and we only have 3.6. I took nullcontext implementation directly from standard library so it should be solid. However, I didn't have time for any testing so let's wait for the bot to tell us whether everything builds.

**Comment by roskakori** [comment]
@Ovsyanka83 pre-commit seems unhappy:
> pygount/common.py:195:12: N801 class name 'nullcontext' should use CapWords convention

I take it you named it that way for consistency with the inconsistent name of the standard library. These Python core developers. 🙄 

I guess the best solution is to keep the inconsistent name for consistency's sake and mark the declaration it as flake8 exception:
```python
# noqa: N801
```

File: pygount/common.py
Comment (roskakori) [inline pygount/common.py:193]: Add
```python
# noqa: N801
```
to fix CI.

File: pygount/common.py
Comment (roskakori) [inline pygount/common.py:192]: Also check for major version 3.

Maybe something like
```python
if sys.version_info[:2] <= (3, 6): ...
```

Or maybe handle it at the import?
```python
try:
    from contextlib import nullcontext
except ImportError:
   # Fallback for Python 3.6.
   class nullcontext(...):
       ...
```

**Review by roskakori — CHANGES_REQUESTED** [review]

File: pygount/common.py
Comment (Ovsyanka83) [reply pygount/common.py:193]: Will do once I'm near my laptop

**Comment by Ovsyanka83** [comment]
I feel like the name is consistent. If I remember correctly, most if not all standard context managers use snake case. In pep8 examples too.

**Comment by roskakori** [comment]
@Ovsyanka83 Hm, pre-commit still fails. Did you activate it as described in [project setup](https://pygount.readthedocs.io/en/latest/contributing.html)?

Regardless, if you just run
```
poetry run pre-commit run --all-files
```
it should fix some formatting issues. You can then commit the resulting changes.

**Comment by Ovsyanka83** [comment]
One of the reasons the build kept failing might be isort. You have configured it and included in pre commit hook but you forgot to add it as a dev dependency. 

The other reason is that my current laptop hates pyenv :)

File: pyproject.toml
Comment (roskakori) [inline pyproject.toml:82]: This should not be needed as the pre-commit hook should provide the dependency. We also don't have poetry dependencies for black, flake8 but pre-commit hooks for them.

**Review by roskakori — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (Ovsyanka83) [reply pyproject.toml:82]: Okay. Give me just a few minutes to remove it

Comment (Ovsyanka83) [reply pyproject.toml:82]: Though doesn't it make sense to include them as dev dependencies? They are in fact dev dependencies and could be used before the hooks.

**Comment by roskakori** [comment]
> One of the reasons the build kept failing might be isort. You have configured it and included in pre commit hook but you forgot to add it as a dev dependency.
> (...)
> P.S. After re-running the pre-commit hook I can say with certainty: isort was the issue.

See comment above.

The pre-commit check during CI runs successfully now, I believe it still should without the isort dependency.

> Also, I am not sure why you include the poetry.lock file. It might be a bad habit but I don't have enough experience with lock files to say anything definitively.

From e.g. [Dependency Management With Python Poetry](https://realpython.com/dependency-management-python-poetry/#distinguish-between-pyprojecttoml-and-poetrylock):

> While the version requirement in the pyproject.toml file can be loose, Poetry locks the versions you’re actually using in the poetry.lock file. That’s why you should commit this file if you’re using Git. By providing a poetry.lock file in a Git repository, you ensure that all developers will use identical versions of required packages.

In particular, I want the CI to use the same versions as I did locally. The requirements for the wheel are derived from the toml, not the lock, so they are more flexible.

**Review by roskakori — APPROVED** [review]
Thanks!

**Pull request merged**
